charm/kafka/reactive/kafka.py

```diff
@@ -13,139 +13,65 @@
-
-def install_snap(upgrade=False):
-    # Need to install the core snap explicit. If not, there's
-    # no slots for removable-media on a bionic install.
-    # Not sure if that's a snapd bug or intended behavior.
-    check_call(['snap', 'install', 'core'])
 
-    cfg = hookenv.config()
-    # Kafka's snap presedence is:
-    # 1. Included in snap
-    # 2. Included with resource of charm of 'kafka'
-    # 3. Snap store with release channel specified in config
-    snap_file = get_snap_file_from_charm() or hookenv.resource_get('kafka')
-    if snap_file:
-        hookenv.log('Detected Kafka snap, installing/refreshing {}'.format(snap_file))
-        check_call(['snap', 'install', '--dangerous', snap_file])
-    if not snap_file:
-        hookenv.log('No Kafka snap detected in charm, installing from snapstore with channel {}'.format(cfg['kafka-release-channel']))
-        
-        if upgrade:
-            check_call(['snap', 'refresh', '--{}'.format(cfg['kafka-release-channel']), KAFKA_SNAP])
-        else:
-            check_call(['snap', 'install', '--{}'.format(cfg['kafka-release-channel']), KAFKA_SNAP])
-
-    # Disable the zookeeper daemon included in the snap, only run kafka
-    check_call(['systemctl', 'disable', 'snap.{}.zookeeper.service'.format(KAFKA_SNAP)])
-    set_state('kafka.available')
-
-
-def get_snap_file_from_charm():
-    snap_files = sorted(glob.glob(os.path.join(hookenv.charm_dir(), "{}*.snap".format(KAFKA_SNAP))))[::-1]
-    if not snap_files:
-        return None
-    return snap_files[0]
+from charms.reactive import (set_flag, when, when_not, hook,
+                             remove_state, set_state)
+from charms.reactive.helpers import data_changed
 
 
-@when('kafka.available')
-@when_not('zookeeper.joined')
-def waiting_for_zookeeper():
-    hookenv.status_set('blocked', 'waiting for relation to zookeeper')
+@when('snap.installed.kafka')
+@when_not('kafka.zk.disabled')
+def disable_bootstrap_zk():
+    check_call([
+        'systemctl', 'disable',
+        'snap.{}.zookeeper.service'.format(KAFKA_SNAP)
+    ])
+    set_state('kafka.zk.disabled')
 
 
-@when('kafka.available', 'zookeeper.joined')
-@when_not('kafka.started', 'zookeeper.ready')
-def waiting_for_zookeeper_ready(zk):
-    hookenv.status_set('waiting', 'waiting for zookeeper to become ready')
+@hook('upgrade-charm')
+def upgrade_charm():
+    remove_state('kafka.zk.disabled')
 
 
-@when_not(
-    'kafka.ca.keystore.saved',
-    'kafka.server.keystore.saved'
-)
-@when('kafka.available')
-def waiting_for_certificates():
-    hookenv.status_set('waiting', 'waiting for easyrsa relation')
+@hook('config-changed')
+def config_changed():
+    remove_state('kafka.configured')
+    set_flag('kafka.force-reconfigure')
 
 
 @when(
-    'kafka.available',
+    'snap.installed.kafka',
     'zookeeper.ready',
     'kafka.ca.keystore.saved',
     'kafka.server.keystore.saved'
 )
 @when_not('kafka.started')
 def configure_kafka(zk):
-    if len(zk.zookeepers()) < 1:
```

Comment:
This status was useful, please revert.

Reply:
@cmars if there are no zk nodes it would go back to waiting for zookeeper relation message, so it’s a bit useless.

Reply:
Ok, I agree.